### DISCUSSION TITLE
Creating MapRoulette mission from a Footprint project from MapSwipe

### DISCUSSION THREAD

**Discussion by frozenhelium** [mapswipe/python-mapswipe-workers#894] [Ideas]
## Get the data from MapSwipe

- Goto [Data](https://mapswipe.org/en/data/) page in the Mapswipe Website
- Find the project you're looking for. (You can use search and filter functionality)
- Go to the Project details page by clicking on the project
- In the **Download the data** section, you can find the necessary data under **Aggregated Results (with Geometry)**
- Download the data using the **Download** button. You'll get a gzipped GeoJSON file (.geojson.gz)
- You can unzip the downloaded data to get the GeoJSON file.
- The data structure of the GeoJSON should look like:

```jsonc
{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "properties": {
        "0_count": 0,
        "1_count": 4,
        "2_count": 0,
        "3_count": 0,
        "total_count": 4,
        "0_share": 0,
        "1_share": 1,
        "2_share": 0,
        "3_share": 0,
        "agreement": 1
        // ...other properties are hidden
      },
      "geometry": {
        // ...properties are hidden
      }
    }
  ]
  // ...other properties are hidden
}
```
<table>
  <tr>
    <td>
      <figure>
        <img alt="mapswipe-data-page" src="https://github.com/mapswipe/python-mapswipe-workers/assets/4931541/5a1d695c-94e7-4b4e-a933-5d12d2397d87" />
        <figcaption>
          MapSwipe Data Page
        </figcaption>
      </figure>
    </td>
    <td>
      <figure>
        <img alt="download-section" src="https://github.com/mapswipe/python-mapswipe-workers/assets/4931541/9e52f827-0e78-4242-b77b-729dfdcd2e90" />
        <figcaption>
          Project Download Section
        </figcaption>
      </figure>
    </td>
  </tr>
</table>

### Understanding the feature properties

The properties in the feature are generated from the responses of the MapSwipe app users of the project.

A footprint project typically has 3 options that app user can choose from.
- Yes (Represented by 1)
- No (Represented by 0)
- Not Sure (Represented by 2)

However, these options can be customized while creating the project. The managers can customize both the label and the value.

Optionally, we can have sub-options for each option.

We have 4 different type of properties that we're interested
#### `<value>_count`
These type of properties represent the number of users who responded with `<value>`
For exampe:
`"0_count": 0` means that 0 users selected **No** for this feature
`"1_count": 4` means that 4 users selected **Yes** for this feature

#### `total_count`

This property represents the total number of users who responded

#### `<value>_share`

These type of properties represent the ratio of "Number of users who responded with `<value>`" to "Total number of users who responded"
i.e. `"0_share" = "0_count" / "total_count"`

#### `agreement`

This property is a statistical measure for assessing the reliability of agreement between a fixed number of raters when assigning categorical ratings to a number of items or classifying items.
More details can be found [here](https://en.wikipedia.org/wiki/Fleiss%27_kappa)

<table>
  <tr>
    <td>
      <img alt="Footprint Options" src="https://github.com/mapswipe/python-mapswipe-workers/assets/4931541/4173017c-b5b5-4126-b0c4-cb101b1916ce" />
      <figcaption>
        Footprint Options
      </figcaption>
    </td>
    <td>
      <img alt="Footprint Suboptions" src="https://github.com/mapswipe/python-mapswipe-workers/assets/4931541/1e498c06-617a-4ff1-bd90-99772acccb35" />
      <figcaption>
        Footprint Suboptions
      </figcaption>
    </td>
  </tr>
</table>

### Filtering

The export contains all the features initially selected while creating the project. But, we are only interested in a subset of the features. To extract only the features of our interest, we can filter the features using the above mentioned properties.

For example: Consider a situation where we want dataset that contains all results where at least 35% of users submitted a **Yes** or **Not Sure** classification.

```js
const geoJson = await loadFromFile('./agg_results_project_id_geom.geojson');
const filteredGeoJson = {
  ...geoJson,
  features: geoJson.features.filter(
    (feature) => {
      const {
        properties: {
          1_share,
          2_share,
        }
      } = feature;

      return (1_share + 2_share) > 0.35;
    }
  ),
}
```

### Grouping

The export might contain a lot of features even after the filtering. It is good to group the features so that we have lesser numer of tasks in MapRoulette.

There are various methods of grouping:
1. Clustering features by proximity
2. Splitting the area into tiles

We can use GIS software to implement the grouping.

## Create a MapRoulette Challenge

Once you have the GeoJSON after the filtering and grouping, you're now ready to create a mission in MapRoullete.

### Upload the GeoJSON

Go to **Create & Manage** page in MapRoulette and open a project where you want to create the challenge (mission).

Click on **Add Challenge** and fill in the details.

Select **I want to upload a GeoJSON file** under **Location of your task data**. After that, you should be able to upload the GeoJSON
Note: If you get the error saying
`GeoJSON error: Polygons and MultiPolygons should follow the right hand rule`
you can use [`geojson-rewind`](https://github.com/mapbox/geojson-rewind) to fix the issue

Fill in the other details as well and click on **Finish** button to complete the process.

It will start building the tasks for the challenge. Once completed, you can proceed to next step

<table>
  <tr>
    <td>
      <figure>
        <img width="500" alt="image7" src="https://github.com/mapswipe/python-mapswipe-workers/assets/4931541/bf0afb47-a78e-4c8f-949f-f62ac944eccc" />
        <figcaption>
          Navigating to "Create & Manage" page
        </figcaption>
      </figure>
    </td>
    <td>
      <figure>
        <img width="500" alt="image5" src="https://github.com/mapswipe/python-mapswipe-workers/assets/4931541/64331fb3-7726-46b8-abb7-c17844e9738f" />
        <figcaption>
          Adding a Challenge
        </figcaption>
      </figure>
    </td>
  </tr>
  <tr>
    <td>
      <figure>
        <img width="500" alt="image2" src="https://github.com/mapswipe/python-mapswipe-workers/assets/4931541/c244c7dd-1d24-497e-a1ec-bd6abe51a325" />
        <figcaption>
          Adding the GeoJSON
        </figcaption>
      </figure>
    </td>
    <td>
      <figure>
        <img width="500" alt="image6" src="https://github.com/mapswipe/python-mapswipe-workers/assets/4931541/3a548213-62fb-4010-b413-f861e9808714" />
        <figcaption>
          Task building progress
        </figcaption>
      </figure>
    </td>
  </tr>
</table>

## Start the MapRoulette Challenge

- Open the project you want to map and click on **Start Challenge**
- Click **Edit** button in the **Completion** section. You should be redirected to OpenStreenMap iD editor
- Select the imagery used for the project from **Background**
- Check if the feature in the area are correctly mapped
- Correct or delete the feature if needed
- Save and upload your edits to OpenStreetMap
- If you made any correction or deleted any feature, please mark **I fixed it!**
  - A pop-up will appear. Describe what was wrong with the buildings you corrected in the comment box of the pop-up
- If no correction or deletion was needed, please mark as **Not an issue**

<table>
  <tr>
    <td>
      <figure>
        <img width="500" alt="image4" src="https://github.com/mapswipe/python-mapswipe-workers/assets/4931541/fa1dc4c3-3f93-4f58-a53e-ca02e367af97" />
        <figcaption>
          Starting a new Challenge
        </figcaption>
      </figure>
    </td>
    <td>
      <figure>
        <img width="500" alt="image3" src="https://github.com/mapswipe/python-mapswipe-workers/assets/4931541/4c852555-8f7b-4457-a335-fece37de1a1c" />
        <figcaption>
          Editing in iD editor
        </figcaption>
      </figure>
    </td>
  </tr>
  <tr>
    <td>
      <figure>
        <img width="500" alt="image1" src="https://github.com/mapswipe/python-mapswipe-workers/assets/4931541/fe712a1c-d8d1-4e64-9070-3785249adf29">
        <figcaption>
          Setting the background imagery
        </figcaption>
      </figure>
    </td>
  </tr>
</table>

**Comment by Jorieke** [comment]
Great write-up!
